### ISSUE TITLE
Loading Chinese front is too slow on opengl2 & glfw2.

### ISSUE THREAD

**Issue by kekxv** [ocornut/imgui#5811]
`ImGui_ImplOpenGL2_NewFrame()` takes more than 30 seconds When I load the Chinese font.

load front code 
```c++
ImFont *ImGui::load_system_font(const std::string &font_name, float front_size) {
    std::string path;
    std::string font_path;
    std::vector<std::string> files;
    if(!font_name.empty()) {
        tools::get_files(tools::get_local_path(), files, font_name);
        if (!files.empty()) {
            font_path = files[0];
        }
    }
    if (font_name.empty() || !tools::exists(font_path)) {
#if defined(WIN32)
        char winpath[MAX_PATH+3];
        memset(winpath,0x00,MAX_PATH+3);
        GetWindowsDirectory(winpath, sizeof(winpath));
        path = tools::make_path(winpath,"Fonts");
#elif defined(__APPLE__)
        path = "/System/Library/Fonts";
#else
        path = "/usr/share/fonts";
#endif
        if (!tools::exists(path)) {
            return nullptr;
        }
        std::string fonts[] = {
                "PingFang.ttc",
                "DejaVuSans.ttf",
                "STHeiti Light.ttc",

                "STSONG.TTF",
                "STXINWEI.TTF",
                "STFANGSO.TTF",
                "STXINGKA.TTF",

                "msyhl.ttc",
                "msjh.ttc",
                "msyh.ttf",

                // "simsun.ttc",
                // "simfang.ttf",
        };
        for (const auto &item: fonts) {
            font_path = tools::find_exists_file(item, path);
            if (!font_path.empty()) {
                break;
            }
        }
    }
    tools::push_info("load font path : " + font_path);
    if (font_path.empty()) {
        return nullptr;
    }

    // Load Fonts
    ImGuiIO &io = ImGui::GetIO();
    // 加载并支持中文字体
    return io.Fonts->AddFontFromFileTTF(font_path.c_str(), front_size, nullptr,
                                        // io.Fonts->GetGlyphRangesChineseSimplifiedCommon());
                                        io.Fonts->GetGlyphRangesChineseFull());
}
```

**Comment by thedemons** [comment]
I don't believe this has anything to do with the backends. The reason it stuck at `ImGui_ImplOpenGL2_NewFrame` is due to the font hasn't been built yet so it was built on demand.

You can build it before entering the render loop using this, to see if the problem really is because of building the font:
```cpp
io.Fonts->Build();
```

Dear ImGui has two official font rasterizers, TrueType (default option) and [FreeType](https://github.com/ocornut/imgui/blob/master/docs/FONTS.md#using-freetype-rasterizer-imgui_freetype), this is what responsible for building the fonts, assuming you're using the TrueType rasterizer, you can measure the performance of the [ImFontAtlasBuildWithStbTruetype](https://github.com/ocornut/imgui/blob/f6bd7c8e53b541dfd6ab22d2684b7c8005a579e4/imgui_draw.cpp#L2335) function to make sure that's the problem.

If it is really that slow, what you could do is build once and write the atlas texture into an image, use it as a cache then load the atlas from it. It requires a lot of code and modifications so I'm not sure if I want to post it here.

**Comment by ocornut** [comment]
Its going to be slow if you load full ranges.
Until we have dynamic font atlas loading the only workarounds are: reducing OversampleH to 1 to divide the time by three, use a custom glyph range using ImGuiGlyphRangeBuilder.

See https://github.com/ocornut/imgui/blob/master/docs/FONTS.md for references.

**Comment by thedemons** [comment]
@ocornut Also, why isn't there an option to use a cached atlas instead of building it every time making the load time longer? Is there any reason not to do so?

**Comment by rorschach-py** [comment]
GetGlyphRangesChineseFull is slow, GetGlyphRangesChineseSimplifiedCommon lacks lots of common chars
try this:
```
// make sure source file is utf8
// add extra chinese char, search: 通用规范汉字表 8105字  
// or load it from txt file
const char * font_char = "一丁七万丈三上下不与丐丑专且丕世丘丙业……";

ImVector<ImWchar> ranges;
ImFontGlyphRangesBuilder builder;

builder.AddRanges(io.Fonts->GetGlyphRangesChineseSimplifiedCommon());
builder.AddText(font_char);
builder.BuildRanges(&ranges);
    
io.Fonts->AddFontFromFileTTF(CFG.font, CFG.fontsize,      NULL, ranges.Data);

```

**Comment by kekxv** [comment]
using `PingFang.ttc` is fast.


slow front:
- msyhl.ttc
- simsun.ttc

**Comment by ocornut** [comment]
FYI, work on `dynamic_fonts` branch aimed to be merged in 1.92 (#8465) solves this, as glyphs are rasterized and packed on demand when they are used.
